File: Session12/C++/AllNodesDistanceKInBinaryTree.cpp
https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/

class Solution {  
    
    void buildParents(unordered_map<TreeNode*, TreeNode*>& parents, TreeNode* root)
    {
        if(root==NULL) {
            return;
        }
        if(root->left!=NULL) {
            parents[root->left] = root;
            buildParents(parents, root->left);
        }
        if(root->right!=NULL) {
            parents[root->right] = root;
            buildParents(parents, root->right);
        }
    }
    
public:
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        
        unordered_map<TreeNode*, TreeNode*> parents;
        buildParents(parents, root);
        
        unordered_set<TreeNode*> visited;
        queue<TreeNode*> q;
        q.push(target);
        vector<int> res;
        
        while(!q.empty() && k>=0) {
            int levelSize = q.size();
            while(levelSize--) {
                TreeNode* curr = q.front();
                q.pop();
                visited.insert(curr);
                
                if(k==0) {
                    res.push_back(curr->val);
                    continue;
                }
                
                // check for left and right child which are not visited
                if(curr->left && !visited.count(curr->left)) {
                    q.push(curr->left);
                }
                if(curr->right && !visited.count(curr->right)) {
                    q.push(curr->right);
                }
                
                // check for unvisited parent
                if(parents.count(curr) && !visited.count(parents[curr])) {
                    q.push(parents[curr]);
                }
            }
            k--;
        }
        return res;
    }
};

File: Session12/Java/AllNodesDistanceKInBinaryTree.java
https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/

class Solution {
    
    void buildParents(Map<TreeNode, TreeNode> hm, TreeNode root)
    {
        if(root == null) return;
        if(root.left != null)
            hm.put(root.left, root);
        
        if(root.right != null)
            hm.put(root.right, root);
    
        buildParents(hm, root.left);
        buildParents(hm, root.right);
        return;
    }
    
    public List<Integer> distanceK(TreeNode root, TreeNode target, int K) {
        
        Map<TreeNode, TreeNode> parentMap = new HashMap<>();
        buildParents(parentMap, root);

        List<Integer> res= new ArrayList<>();
        Set<TreeNode> visited = new HashSet<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(target);
        
        while(!q.isEmpty() && K>=0){
            int size= q.size();
            for(int i =0; i <size; i++){
                TreeNode curr = q.poll();
                visited.add(curr);
                if(K == 0){
                  res.add(curr.val);
                  continue;
                }
                if(parentMap.containsKey(curr) && !visited.contains(parentMap.get(curr))){
                    q.add(parentMap.get(curr));
                }
                if(curr.left != null && !visited.contains(curr.left)){
                    q.add(curr.left);      
                }
                if(curr.right != null && !visited.contains(curr.right)){
                    q.add(curr.right);      
                }
            }
            K--;
        }
        return res;
    }
}